xdsl/dialects/onnx.py

```diff
@@ -254,6 +254,105 @@ def verify_(self) -> None:
+            attributes={"allowzero": allow_zero},
+            operands=[data, shape],
+            result_types=[data.type],
+        )
+
+    def verify_(self) -> None:
+        if (
+            not isinstance(data_type := self.data.type, TensorType)
+            or not isinstance(shape_type := self.shape.type, TensorType)
+            or not isinstance(reshaped_type := self.reshaped.type, TensorType)
+        ):
+            assert (
+                False
+            ), "onnx elementwise operation operands and result must be of type TensorType"
+
+        data_type = cast(TensorType[Attribute], data_type)
+        reshaped_type = cast(TensorType[Attribute], reshaped_type)
+        shape_type = cast(TensorType[Attribute], shape_type)
+
+        if shape_type.element_type != IntegerType(64):
+            raise VerifyException(
+                "shape element type has to be a 64-bit signless integer"
+            )
+
+        data_type = data_type.get_shape()
+        shape_type = shape_type.get_shape()
+        reshaped_type = reshaped_type.get_shape()
+
+        # Shape tensor rank can't be -1
+        if shape_type[0] == -1:
+            raise VerifyException("Shape tensor rank must not be equal to -1")
+
+        # There is currently only support for rank one shape tensors in onnx-mlir
+        # Shape tensor must have a constant shape
+        if len(shape_type) != 1:
+            raise VerifyException("Shape tensor must have a rank one")
+
+        # The input tensor's shape and the output tensor's shape are required to have the same number of elements.
+        if len(data_type) != len(reshaped_type):
+            raise VerifyException(
+                "Input tensor's shape and output tensor's shape must have the same number of elements"
+            )
+
+        # At most one dimension of the new shape can be -1.
+        # In this case, the value is inferred from the size of the tensor and the remaining dimensions.
+        new_shape = list(reshaped_type)
+        count_minus_one = new_shape.count(-1)
+        if count_minus_one == 1:
+            index_of_minus_one = new_shape.index(-1)
+            specified_dim = len(new_shape)
+            total_elements = len(data_type)
+            missing_dim = total_elements // specified_dim
+            new_shape[index_of_minus_one] = missing_dim
+
+        # Handle case where dimension is zero
+        for i, dim in enumerate(new_shape):
+            if dim == 0:
+                if self.allow_zero:
+                    # If allow_zero is set, explicitly set the dimension to zero  (i.e. not taken from input tensor)
+                    new_shape[i] = 0
+                else:
+                    # dimension is 0, leave it unchanged  (i.e. taken from the input tensor).
+                    new_shape[i] = data_type[i]
+
+        # Shape (second input) could be an empty shape, which means converting to a scalar.
+        if len(shape_type) == 0:
+            shape_type = IntegerType(64)
```

Comment:
what does this do?

Reply:
i guess this is code that isn't exactly supposed to be in `verify_`, but the first bit checks if atleast one dim of the reshaped tensor (new shape) is -1, then it infers the actual dimension, the next bit (for loop section) checks the dimension is zero in the new shape according the the conditions in the doc

Reply:
i guess i could put these in helper functions actually

Reply:
it's just that it looks like dead code, this has no effect

Reply:
okay, is it because im not setting it to the inital right

Reply:
this is setting a local variable, not `self`. I would say that `verify_` should never mutate state, this sort of inference may be better done in a rewrite pass.

Reply:
@superlopuh , i've been able to sort the file changes out, but i've been trying to understand why code is dead code in the block you highlighted, if i want to add what i said is that even needed now?

Reply:
I would recommend removing this part of the logic, and then I think it should be ready to merge

Reply:
okay i will remove it, id appreciate (if you have time later) to give me a bit more of an understanding on these sort of inferences, or point me to somewhere to read more on it

Reply:
Sure! Chapter 4 of the MLIR Toy tutprial talks about it a little bit, although it's made a little more complicated than it needs to be by using a lot of abstractions for the implementation of the shape inference: https://mlir.llvm.org/docs/Tutorials/Toy/Ch-4/

The main idea is to be able to represent the IR at all, even with casts and reshapes, and then have a dedicated rewrite step to rewrite it into the cleaner representation once the shape information is available. The ONNX operations we probably want to represent as close to the input as possible, and then use compiler transformations to mutate them, to leverage all the infrastructure we've built for that.